Libplanet/Net/BlockCompletion.cs

```diff
@@ -0,0 +1,335 @@
+                        );
+                    },
+                    yield.CancellationToken
+                );
+            });
+        }
+
+        private void Demand(HashDigest<SHA256> blockHash, bool retry)
+        {
+            if (Satisfies(blockHash, ignoreTransientCompletions: retry))
+            {
+                return;
+            }
+
+            if (_satisfiedBlocks.TryAdd(blockHash, false) || retry)
+            {
+                _demands.Enqueue(blockHash);
+                _started = true;
+                _demandEnqueued.Release();
+                _logger.Verbose("A demand was enqueued: {BlockHash}", blockHash);
+            }
+        }
+
+        private void Demand(IEnumerable<HashDigest<SHA256>> blockHashes, bool retry)
+        {
+            foreach (HashDigest<SHA256> hash in blockHashes)
+            {
+                Demand(hash, retry);
+            }
+        }
+
+        internal class PeerPool
+        {
+            private readonly ConcurrentQueue<TPeer> _completions;
+            private readonly ConcurrentDictionary<TPeer, Task> _tasks;
+            private long _taken;
+            private long _finished;
+
+            public PeerPool(IEnumerable<TPeer> peers)
+            {
+                _completions = new ConcurrentQueue<TPeer>(peers);
+                _tasks = new ConcurrentDictionary<TPeer, Task>();
+                _taken = 0;
+                _finished = 0;
+            }
+
+            public async Task SpawnAsync(
+                Func<TPeer, Task> action,
+                CancellationToken cancellationToken = default
+            )
+            {
+                Interlocked.Increment(ref _taken);
+                TPeer peer;
+                while (!_completions.TryDequeue(out peer))
+                {
+                    Task[] tasks = _tasks.Values
+                        .Concat(new[] { Task.Delay(500, cancellationToken) })
+                        .ToArray();
+                    await Task.WhenAny(tasks);
+                    cancellationToken.ThrowIfCancellationRequested();
+                }
+
+                if (_tasks.TryRemove(peer, out Task completeTask))
+                {
+                    await completeTask;
+                }
+
+                _tasks[peer] = Task.Run(
+                    async () =>
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+                        try
+                        {
+                            await action(peer);
+                        }
+                        finally
+                        {
+                            _completions.Enqueue(peer);
+                            Interlocked.Increment(ref _finished);
+                        }
+                    },
+                    cancellationToken: cancellationToken
+                );
+            }
+
+            public async Task WaitAll(CancellationToken cancellationToken = default)
```

Comment:
```suggestion
            public async Task WaitAll(CancellationToken cancellationToken = default(CancellationToken))
```
Isn't this does not work in some environments?